### ISSUE TITLE
Joining table

### ISSUE THREAD

**Issue by FrancoisM** [Azure/azure-mobile-services#522]
Hi,

How can we join tables? Not being able to do it is an issue. For example, say I have a table images with columns id, res, res2x, res3x where I store all my images with different résolutions. My other tables are related to images with a foreign key image.
Without being able to join I have to do things like

var myItems=await getsynctable<MyTable>().Where(x=>some condition).ToListAsync();
var list = myObjs.select(x=>x.Image).ToList();
var img=getsyncTable<Images>().Where(x=>list.Contains(x.Id)).ToList();
var result=new MyClass();
foreach(var item in myItems)
{
    result.MyItem=item;
    result.MyImage=img.Where(x=>x.Id==item.Image);
}
return result;

This is heavy and had caused me some parse overflow when the list is big enough (as contains gets transformed into id==a || id==b || id==c ....................)
I could also do something that looks simpler but which queries the DB multiple times rather than once and hence has some impact on performance:

class MyItem
{
    .....
    string MyImage
    {
        get
        {
              return await client.GetSyncTable<Images>().Where(x=>x.Id==Image).SingleOrDefault<string>();
        }
    }
}

Conclusion: how can I simply join? :-)

**Comment by Rhywden** [comment]
Uh, could you provide an example of what you mean?

Because I am currently having the same problem of needing Joins and not having them. In a many-to-many relationship, no less.

**Comment by ltaljaard** [comment]
Hello lindydonna. In my app I have all kinds of relationships: 1:1, 1:m and m:n. I would like to make use of GetSyncTable() as it does incremental sync's. What would it take to make this work? And why are you not looking to support relationships? To me this seems like a fundamental shortcoming as virtually all databases have relationships in them. If you could perhaps advise the best way forward for my app in this regard?

Then a second (related) question please. Up to this point, I have at least managed to pull all the related data that I'm after from the backend DB into the client app with a single call using ReadAsync() - using a combination of OData $expand and $filter parameters, please see the code snippet below:

var agent_json = await mobileServiceClient.GetTable<Agent>().ReadAsync("$expand=Person,Roles,Agency,Agency/Candidates,Agency/Candidates/Person,&$filter=Person/LogonId eq '" + wa_userID + "'");	
Agent agent_with_related_entities = Newtonsoft.Json.JsonConvert.DeserializeObject<Agent>(agents_json.First.ToString());

That fills my agent_with_related_entities nicely with all the nested objects, so that part of it seems like a solution for fetching the data (filtered according to what the local DB may receive) at least.

So now I have all the data that I'd like to seed my local device's SQLite database with. How can I write this data into the local database? Will something like this work perhaps: mobileServiceClient.GetSyncTable<Agent>().UpdateAsync(agent_with_related_entities);

How will the above UpdateAsync() handle the nested objects, will it save them automatically? Or will I have to manually iterate over them all and call UpdateAsync() on each?

In summary, there are two parts to my question:
1. How to (best) initially seed the local DB
2. How to keep the local DB in sync with the server when it comes to the RELATIONSHIPS - 1:1, 1:m and m:n relationships could all change on the server-side, and then the local DB must be updated accordingly - preferably using an incremental sync approach as is available in GetSyncTable(), but if that is not an option I would need some advice as to how to implement such functionality myself (under tight time constraints so hopefully something that won't involve open brain surgery?)

**Comment by ltaljaard** [comment]
Hello, can I please have your opinion on the following workaround for the current lack of support for relationships in Azure Mobile Services?

At app startup on the mobile device, do the following query to the server to read all the ID's only of the related objects to the given Agent object that we're trying to pull into the mobile device's DB:

var agents_json = await mobileServiceClient.GetTable<Agent>().ReadAsync(					"$expand=Person,Roles,Agency,Agency/Candidates,Agency/Candidates/Person" +
"&$select=Person/Id,Roles/Id,Agency/Id,Agency/Candidates/Id,Agency/Candidates/Person/Id" +
"&$filter=Person/LogonId eq '" + wa_userID + "'");	
Agent agent_with_related_entities = Newtonsoft.Json.JsonConvert.DeserializeObject(agents_json.First.ToString());

Then loop through the returned object graph and call PullAsync() for each related object as follows - e.g. for Person:

mobileServiceClient.GetSyncTable<Person> ().PullAsync("PersonWithID" + [person ID obtained from traversing the above graph], mobileServiceClient.GetSyncTable<Person> ().Where(p => p.LogonId == [person ID obtained from traversing the above graph]));

Would that effectively give me incremental sync for all the related objects? And would that also "soft delete" (assuming that soft delete is turned on) any objects in the local device DB (and/or their relationships) that are NOT related anymore (i.e. where the related object was deleted on the server or where the relationship between the two objects were broken) ?

**Comment by biomorph00** [comment]
Hello lindydonna.   Can you please let us know if relationship support is on the roadmap yet?
Our company is currently evaluating mobile platforms and I think this will be a deal-breaker for us.
It seems ridiculous to us that new feature after new feature is continually pumped out without getting the basics sorted.  I expected better from Microsoft.
Currently I think our company will be going with Kony which looks so limited to me, but after a lot of people have been trying for weeks to get a proper CRUD working easily with AMS/html client and the lack of support for relationships has totally soured the experience.

**Comment by paulbatum** [comment]
We have received many requests for relationship support. We need to start consolidating this feedback into a single feature writeup. See https://github.com/Azure/azure-mobile-apps/issues/6.
